### PR TITLE
Add /NOIMPLIB and /NOEXP when using MSVC and not clang.

### PR DESCRIPTION
**Pull request by dave-juicelabs**

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -146,11 +146,13 @@ if vkd3d_platform == 'windows'
   if vkd3d_compiler.has_link_argument('-static-libstdc++')
     add_global_link_arguments('-static-libstdc++', language : [ 'cpp' ])
   endif
-  if vkd3d_compiler.has_link_argument('/NOIMPLIB')
-    add_global_link_arguments('/NOIMPLIB', language : [ 'c', 'cpp' ])
-  endif
-  if vkd3d_compiler.has_link_argument('/NOEXP')
-    add_global_link_arguments('/NOEXP', language : [ 'c', 'cpp' ])
+  if not vkd3d_is_msvc
+    if vkd3d_compiler.has_link_argument('/NOIMPLIB')
```
Comment (HansKristian-Work) [inline meson.build:150]: Isn't this doing the opposite of that what MR text is claiming?

Comment (HansKristian-Work) [reply meson.build:150]: Also, why should this not be run on clang? We really don't want to export from the .exe.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
question

**Comment by ishitatsuyuki** [comment]
Breaking up flags into GNU-like and MSVC-like (including clang-cl) makes sense, as clang-cl would take both type of options (see [doitsujin/dxvk@`f0ecf6d` (#3808)](https://github.com/doitsujin/dxvk/pull/3808/commits/f0ecf6ddb087468aa5d381b96f617a9bf1cbaaae)). But as Hans-Kristian said, this PR seems to be doing the opposite of the claim.

File: meson.build (hunk 1 above)
Comment (dave-juicelabs) [reply meson.build:150]: This is the error received when building with clang on Windows
```
[57/101] Linking target libs/d3d12/d3d12.dll
FAILED: libs/d3d12/d3d12.dll
"lld-link"  /MACHINE:x64 /OUT:libs/d3d12/d3d12.dll libs/d3d12/d3d12.dll.p/main.c.obj "/nologo" "/OPT:REF" "/DLL" "/DEF:..\..\libs\d3d12\d3d12.def" "/IMPLIB:libs\d3d12\d3d12.lib" "/NOIMPLIB" "/NOEXP" "libs\vkd3d-common\libvkd3d_common.a" "dxgi.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
lld-link: error: could not open '/NOEXP': no such file or directory
```

Comment (ishitatsuyuki) [reply meson.build:150]: What does your `meson-logs/meson-log.txt` say? It should detect `/NOEXP` as unsupported.

Comment (dave-juicelabs) [reply meson.build:152]: Force pushed an update resolving the msvc vs clang issue. vkd3d_is_msvc was checking both msvc and clang-cl so the if statement still worked but would have broken msvc. I built with both msvc and clang-cl without an issue with these changes.

Comment (ishitatsuyuki) [reply meson.build:152]: Again it should be covered by the link flag check though. On clang-cl on Linux with msvc-wine this is what I get:
```
Cached command line:  clang-cl -fuse-ld=lld /EHsc /winsysroot /home/ishitatsuyuki/msvc /home/ishita>

Code:
 int main(void) { return 0; }

Cached compiler stdout:
 
Cached compiler stderr:
 lld-link: error: could not open '/NOEXP': No such file or directory
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)

Compiler for C supports link arguments /NOEXP: NO (cached)
```

Comment (dave-juicelabs) [reply meson.build:152]: Sorry, missed your comment above about the meson-log.txt.
```
Running compile:
Working directory:  D:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpxjob0ahz
Command line:  clang-cl D:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpxjob0ahz\testfile.c /FeD:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpxjob0ahz\output.exe /nologo /showIncludes /utf-8 /MDd /nologo /showIncludes /utf-8 /Od /Oi- /link /NOIMPLIB 

Code:
 int main(void) { return 0; }

Compiler stdout:
 
Compiler stderr:
 
Compiler for C supports link arguments /NOIMPLIB: YES 
Running compile:
Working directory:  D:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpiorikpdw
Command line:  clang-cl D:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpiorikpdw\testfile.c /FeD:\Github\juicelabs\juice\client\vkd3d\build\debug\meson-private\tmpiorikpdw\output.exe /nologo /showIncludes /utf-8 /MDd /nologo /showIncludes /utf-8 /Od /Oi- /link /NOEXP 

Code:
 int main(void) { return 0; }

Compiler stdout:
 
Compiler stderr:
 
Compiler for C supports link arguments /NOEXP: YES 
```

So, it says it is supported however the build produces
```
[1/32] Linking target libs/d3d12/d3d12.dll
FAILED: libs/d3d12/d3d12.dll libs/d3d12/d3d12.pdb
"lld-link"  /MACHINE:x64 /OUT:libs/d3d12/d3d12.dll libs/d3d12/d3d12.dll.p/main.c.obj "/nologo" "/DEBUG" "/PDB:libs\d3d12\d3d12.pdb" "/DLL" "/DEF:..\..\libs\d3d12\d3d12.def" "/IMPLIB:libs\d3d12\d3d12.lib" "/NOIMPLIB" "/NOEXP" "libs\vkd3d-common\libvkd3d_common.a" "dxgi.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
lld-link: error: could not open '/NOEXP': no such file or directory
[2/32] Linking target tests/descriptor-performance.exe
FAILED: tests/descriptor-performance.exe tests/descriptor-performance.pdb
"lld-link"  /MACHINE:x64 /OUT:tests/descriptor-performance.exe tests/descriptor-performance.exe.p/descriptor_performance.c.obj "/nologo" "/DEBUG" "/PDB:tests\descriptor-performance.pdb" "/NOIMPLIB" "/NOEXP" "tests\libd3d12-test-utils.a" "libs\vkd3d-common\libvkd3d_common.a" "d3d12.lib" "dxgi.lib" "d3d12.lib" "dxgi.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
lld-link: error: could not open '/NOEXP': no such file or directory
[3/32] Linking target tests/pso-library-bloat.exe
FAILED: tests/pso-library-bloat.exe tests/pso-library-bloat.pdb
"lld-link"  /MACHINE:x64 /OUT:tests/pso-library-bloat.exe tests/pso-library-bloat.exe.p/pso_library_bloat.c.obj "/nologo" "/DEBUG" "/PDB:tests\pso-library-bloat.pdb" "/NOIMPLIB" "/NOEXP" "tests\libd3d12-test-utils.a" "libs\vkd3d-common\libvkd3d_common.a" "d3d12.lib" "dxgi.lib" "d3d12.lib" "dxgi.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
lld-link: error: could not open '/NOEXP': no such file or directory
[4/32] Linking target tests/shader-api.exe
FAILED: tests/shader-api.exe tests/shader-api.pdb
"lld-link"  /MACHINE:x64 /OUT:tests/shader-api.exe tests/shader-api.exe.p/vkd3d_shader_api.c.obj "/nologo" "/DEBUG" "/PDB:tests\shader-api.pdb" "/NOIMPLIB" "/NOEXP" "tests\libd3d12-test-utils.a" "libs\vkd3d-common\libvkd3d_common.a" "libs\vkd3d-shader\libvkd3d-shader.a" "subprojects\dxil-spirv\libdxil-spirv.a" "d3d12.lib" "dxgi.lib" "d3d12.lib" "dxgi.lib" "/SUBSYSTEM:CONSOLE" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
lld-link: error: could not open '/NOEXP': no such file or directory
```

Comment (dave-juicelabs) [reply meson.build:152]: May be a version issue
```
clang version 15.0.1
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\x64\bin
```

Comment (ishitatsuyuki) [reply meson.build]: Hm, I suppose there's a bug where Meson detects the linker as lld-link while the test compile uses link.exe...

**Pull request merged**